File: kotlin/workers/src/io/bazel/ruleskotlin/workers/Context.java
/*
 * Copyright 2018 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.bazel.ruleskotlin.workers;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Context {
    private final EnumMap<Flags, String> args = new EnumMap<>(Flags.class);
    private final Map<Meta<?>, Object> meta = new HashMap<>();

    private static final Map<String, Flags> ALL_FIELDS_MAP = Arrays.stream(Flags.values()).collect(Collectors.toMap(x -> x.name, x -> x));
    private static final Flags[] MANDATORY_FIELDS = Arrays.stream(Flags.values()).filter(x -> x.mandatory).toArray(Flags[]::new);

    private Context(List<String> args) {
        if (args.size() % 2 != 0) {
            throw new RuntimeException("args should be k,v pairs");
        }

        for (int i = 0; i < args.size() / 2; i++) {
            String flag = args.get(i * 2);
            String value = args.get((i * 2) + 1);
            Flags field = ALL_FIELDS_MAP.get(flag);
            if (field == null) {
                throw new RuntimeException("unrecognised arg: " + flag);
            }
            this.args.put(field, value);
        }

        for (Flags mandatoryField : MANDATORY_FIELDS) {
            if (!this.args.containsKey(mandatoryField)) {
                throw new RuntimeException("mandatory arg missing: " + mandatoryField.name);
            }
        }
    }

    public static Context from(List<String> args) {
        return new Context(args);
    }

    public EnumMap<Flags, String> of(Flags... fields) {
        EnumMap<Flags, String> result = new EnumMap<>(Flags.class);
        for (Flags field : fields) {
            String value = args.get(field);
            if (value != null) {
                result.put(field, value);
            }
        }
        return result;
    }

    public interface Action extends Consumer<Context> {
    }

    public void apply(Action... consumers) {
        Stream.of(consumers).forEach(c -> c.accept(this));
    }


    String get(Flags field) {
        return args.get(field);
    }

    @SuppressWarnings("unchecked")
    <T> T get(Meta<T> key) {
        return (T) meta.get(key);
    }

    @SuppressWarnings("unchecked")
    <T> T putIfAbsent(Meta<T> key, T value) {
        return (T) meta.putIfAbsent(key, value);
    }
}

File: kotlin/workers/src/io/bazel/ruleskotlin/workers/Context.kt
/*
 * Copyright 2018 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("UNCHECKED_CAST")

package io.bazel.ruleskotlin.workers

import java.util.*
import java.util.stream.Stream

class Context private constructor(args: List<String>) {
    private val args = EnumMap<Flags, String>(Flags::class.java)
    private val meta = HashMap<Meta<*>, Any>()

    init {
        if (args.size % 2 != 0) {
            throw RuntimeException("args should be k,v pairs")
        }

        for (i in 0 until args.size / 2) {
            val flag = args[i * 2]
            val value = args[i * 2 + 1]
            val field = ALL_FIELDS_MAP[flag] ?: throw RuntimeException("unrecognised arg: " + flag)
            this.args[field] = value
        }

        MANDATORY_FIELDS.asSequence()
                .filterNot { this.args.containsKey(it) }
                .forEach { throw RuntimeException("mandatory arg missing: " + it.globalFlag) }
    }

    fun of(vararg fields: Flags): EnumMap<Flags, String> {
        val result = EnumMap<Flags, String>(Flags::class.java)
        for (field in fields) {
            val value = args[field]
            if (value != null) {
                result[field] = value
            }
        }
        return result
    }

    fun apply(vararg consumers: (Context) -> Unit) {
        Stream.of(*consumers).forEach { it(this) }
    }

    internal operator fun get(field: Flags): String? = args[field]
    internal operator fun <T : Any> get(key: Meta<T>): T? = meta[key] as T?
    internal fun <T : Any> putIfAbsent(key: Meta<T>, value: T): T? = meta.putIfAbsent(key, value as Any) as T?

    companion object {
        private val ALL_FIELDS_MAP = Flags.values().map { it.globalFlag to Flags.valueOf(it.name) }.toMap()
        private val MANDATORY_FIELDS = Flags.values().filter { x -> x.mandatory }

        fun from(args: List<String>): Context {
            return Context(args)
        }
    }
}
